tests/Collection/CodecCollectionFunctionalTest.php

```diff
@@ -0,0 +1,530 @@
+    }
+
+    public static function provideBulkWriteOptions(): Generator
+    {
+        $replacedObject = TestObject::createDecodedForFixture(3);
+        $replacedObject->x->foo = 'baz';
+
+        yield 'Default codec' => [
+            'expected' => [
+                TestObject::createDecodedForFixture(1),
+                TestObject::createDecodedForFixture(2),
+                $replacedObject,
+                TestObject::createDecodedForFixture(4),
+            ],
+            'options' => [],
+        ];
+
+        $replacedObject = new BSONDocument(['_id' => 3, 'id' => 3, 'x' => new BSONDocument(['foo' => 'baz']), 'decoded' => false]);
+        $replacedObject->x->foo = 'baz';
+
+        yield 'No codec' => [
+            'expected' => [
+                self::createFixtureResult(1),
+                self::createFixtureResult(2),
+                $replacedObject,
+                self::createObjectFixtureResult(4, true),
+            ],
+            'options' => ['codec' => null],
+        ];
+
+        yield 'BSON type map' => [
+            'expected' => [
+                Document::fromPHP(self::createFixtureResult(1)),
+                Document::fromPHP(self::createFixtureResult(2)),
+                Document::fromPHP($replacedObject),
+                Document::fromPHP(self::createObjectFixtureResult(4, true)),
+            ],
+            'options' => ['typeMap' => ['root' => 'bson']],
+        ];
+    }
+
+    /** @dataProvider provideBulkWriteOptions */
+    public function testBulkWrite($expected, $options): void
+    {
+        $this->createFixtures(3);
+
+        $replaceObject = TestObject::createForFixture(3);
+        $replaceObject->x->foo = 'baz';
+
+        $operations = [
+            ['insertOne' => [TestObject::createForFixture(4)]],
+            ['replaceOne' => [['_id' => 3], $replaceObject]],
+        ];
+
+        $result = $this->collection->bulkWrite($operations, $options);
+
+        $this->assertInstanceOf(BulkWriteResult::class, $result);
+        $this->assertSame(1, $result->getInsertedCount());
+        $this->assertSame(1, $result->getMatchedCount());
+        $this->assertSame(1, $result->getModifiedCount());
+
+        // Extract inserted ID when not using codec as it's an automatically generated ObjectId
+        if ($expected[3] instanceof BSONDocument && $expected[3]->_id === null) {
+            $expected[3]->_id = $result->getInsertedIds()[0];
+        }
+
+        if ($expected[3] instanceof Document && $expected[3]->get('_id') === null) {
+            $inserted = $expected[3]->toPHP();
+            $inserted->_id = $result->getInsertedIds()[0];
+            $expected[3] = Document::fromPHP($inserted);
+        }
+
+        $this->assertEquals(
+            $expected,
+            $this->collection->find([], $options)->toArray(),
+        );
+    }
+
+    public function provideFindOneAndModifyOptions(): Generator
```

Comment:
Consider adding a regression test for the issue I mentioned in https://github.com/mongodb/mongo-php-library/pull/1140/files#r1284677901. You'd need some `findAndModify` options to result in `value` being unset on the response.

See [findAndModify: value](https://www.mongodb.com/docs/manual/reference/command/findAndModify/#value) for two possible conditions to trigger that response.